src/deadline/client/_pid_utils.py

```diff
@@ -1,85 +1,148 @@
         if not os.path.exists(pid_file_full_path):
-            logger.echo(f"Download progress file does not exist at {pid_file_full_path}")
-            # Create a new pid file with the current process id
-            return _obtain_pid_lock_atomically(pid_file_full_path, logger)
-
-        # Try to open pid file at download progress location in read mode
-        with open(pid_file_full_path, "r") as f:
-            # Read pid file and obtain the process id from file contents
-            pid = f.read()
-            try:
-                # Get process using the process id
-                psutil.Process(int(pid))
-                # Process with the pid exists, so we cannot obtain a lock
-                raise RuntimeError(
-                    f"Another download is in progress at {saved_progress_checkpoint_location}, use --force-bootstrap or wait for previous download to finish"
-                )
-            except psutil.NoSuchProcess:
-                # No such process exists with the process id, so we can delete the pid file
-                logger.echo(f"Process with pid {pid} is not running. Deleting pid file.")
-
-                f.close()  # Close the file before over-writing it
-
-                # Create a new pid file with the current process id
-                return _obtain_pid_lock_atomically(pid_file_full_path, logger)
+            print_function_callback(
+                f"Download pid lock file does not exist at {pid_file_full_path}"
+            )
+
+            can_obtain_pid_lock = True
+
+        # 3. Try to read existing pid file to verify process with the pid lock is active & hasn't ended pre-maturely
+        else:
+            with open(pid_file_full_path, "r") as f:
+                pid = f.read()
+                try:
+                    psutil.Process(int(pid))
+                    # Process with the pid exists, so we cannot obtain a lock
+                    print_function_callback(
+                        f"Unable to acquire pid lock as process with pid {pid} exists on {pid_file_full_path}"
+                    )
+                    raise RuntimeError(
+                        f"Unable to acquire pid lock as process with pid {pid} exists on {pid_file_full_path}"
+                    )
+                except psutil.NoSuchProcess:
+                    # No such process exists with the process id, so we can delete the pid file
+                    print_function_callback(
+                        f"Process with pid {pid} is not running. Deleting pid file."
+                    )
+
+                    f.close()  # Required for windows
+
+                    # Delete existing pid file as pid is not active
+                    # Once we've determined the process in pid file isn't running we do nothing except:
+                    # Closing file (required for windows) and deleting it
+                    # This will keep race conditions across concurrent processes trying this to be least possible without locking
```

Comment:
The race starts at the `f.read()` call on line 46 and ends at the `os.remove` on line 69. I see that there's a call to `print_function_callback` in the line of execution, which makes it longer. It's a bit hard to see the execution paths because of the try/except pattern here, so I think switching it to use `psutil.pid_exists` will make it easier to read and see the race. Something like this:

```python
with open(pid_file_full_path, "r") as f:
    pid = f.read()
if not psutil.pid_exists(int(pid)):
    os.remove(pid_file_full_path)
    print_function_callback(...)
else:
    print_function_callback(...)
    raise ...
```